### ISSUE TITLE
Dynamic PGO work planned for .NET 7

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/runtime#64659]
**This issue captures the planned work for .NET 7. This list is expected to change throughout the release cycle according to ongoing planning and discussions, with possible additions and subtractions to the scope.**

## Summary
Dynamic PGO is an opt-in feature introduced in .NET 6. In .NET 7 we will work on making Dynamic PGO easier to use, applicable to more scenarios, and offering more performance improvements. 

As part of this work, we will also enable On-Stack Replacement (OSR) and change the default for QuickJitForLoops for x64 and arm64. This will improve Dynamic PGO by including methods with loops in the Tier0 set. It will also notably improve startup time for apps that have jit-intensive startup interval.

## Planned for .NET 7

- [x] (EgorBo WIP) Add support for enabling Dynamic PGO in SDK projects and runtime configs (#70410, #71438, https://github.com/dotnet/sdk/pull/26350)
- [x] (Aman WIP) Hot/Cold splitting (#69763, #69922, #70708, #71236, #71273) -- and more over in dotnet/runtimelabs; eg https://github.com/dotnet/runtimelab/pull/1900, https://github.com/dotnet/runtimelab/pull/1915, ...
- [x] Add GDV-like capability for delegate calls (https://github.com/dotnet/runtime/pull/68703)
- [x] Clone loops based on presence of loop-invariant GDV tests in the loop body (#65206) https://github.com/dotnet/runtime/pull/70377
- [x] Above depends on: https://github.com/dotnet/runtime/issues/70100 and https://github.com/dotnet/runtime/issues/61040
- [x] Fix OSR x64 epilog unwind (see [note](https://github.com/dotnet/runtime/pull/64116#issuecomment-1022795565))  https://github.com/dotnet/runtime/pull/65609
- [x] Enable OSR by default for x64/arm64 https://github.com/dotnet/runtime/pull/65675
- [x] Address microbenchmark perf issues seen with QJFL=1 and OSR (see [note](https://github.com/dotnet/runtime/pull/61934#issuecomment-990031427)) https://github.com/dotnet/BenchmarkDotNet/pull/1949 and https://github.com/dotnet/BenchmarkDotNet/pull/1935
- [x] #52370
- [x] #55325 (more generally: optimize type tests via PGO)

## Under Consideration
- (EgorBo WIP for Native AOT) Add no-fallback GDV when runtime can provide full set of possible classes
 
## Cut from .NET 7 Plans
- (EgorBo in PR) #70941
- #62457 (too many missing pieces)
- Fix issues in profile maintenance: #50419, #50030, #48476
- #48925
- #46885
- #46883
- Reduce performance overhead of Tier0 instrumentation
- Add GDV-like capability for indirect calls (#44610)
- #46887 (needs SSA update)

**Comment by manofstick** [comment]
Hi!

...Sorry if this isn't the correct place for this post...

I've been playing around with PGO in .net 6 with mostly excellent results, but...

I've hit one case where it appears (to ignorant me) that the statistics collecting method isn't being replaced by the optimized version (well at least that is my simple understanding of what PGO should be doing...)

Anyway I'm not sure what information I can provide, I have tried creating a simple test case (with BenchmarkDotNet) but when it the simple case it appeared to do the "correct thing". i.e. works at the equivalent-ish speed.

But in my "production" case it consistently, without fail, always generates badness.

The method is generated using Expression trees - i.e. compiled Lambda - but is just a very simple object construction based on a `BinaryReader`, equivalent to the following c#:

```
SomeBusinessType ReadItem(BinaryReader reader)
{
    return new SomeBusinessType (
        reader.ReadUInt16(),
        CustomSerializationMethods.Methods.binaryReadCustomType(reader),
        reader.ReadSingle()
    );
}
```

The `SomeBusinessType` is an F# struct, but doesn't appear to be anything out of the ordinary.

It works perfectly fine and performant in "normal" .net 6, or with the PGO environmental variables set and debug build (although I assume that's just because it doesn't engage the PGO engine?).

![image](https://user-images.githubusercontent.com/2093966/162900098-bd5fb2ee-1074-437b-92ec-550c0e4839fb.png)

As you can see in the image from the profiler, the compiled lambda, and a function `JIT_ClassProfile32` completely dominate.

(It's called in a tight loop, populating an `ImmutableArray<>.Builder`, running on multiple threads - on Tasks started with `Task.Run()`. After a successful, albeit slow load, restarting this loading process, from within the same process, it still runs slowly - i.e. still not swapping out the function.)

Anyway, it's in proprietary code, and given that my simplified version didn't exhibit the same issue, I'm not sure what I can offer, but you can direct me to provide further information if you feel the need.

**Comment by AndyAyersMS** [comment]
> ...Sorry if this isn't the correct place for this post...

No worries, this is a good place.

What environment vars are you setting? If you've set `COMPlus_TC_QuickJitForLoops=1` then it's possible for methods to get "trapped" in Tier0 code for long stretches of time, which would be consistent with what you're seeing above.

If so, the general fix for this is [On Stack Replacement](https://github.com/dotnet/runtime/blob/main/docs/design/features/OnStackReplacement.md) (aka OSR) which will be on by default in .NET 7 Preview 4.

There is an unsupported early version of OSR in .NET 6 which may or may not work... feel free give it a try by setting `COMPlus_TC_OnStackReplacement=1`.

You can also mitigate in source if you can restructure the code so that your loop is tiled into a two-level loop where the inner loop is in a helper method and the outer loop runs at least 30 iterations (split 10,000 into say -> 100 x 100). I realize this may not be easy if you don't have explicit looping.

**Comment by manofstick** [comment]
Yes indeed I had `COMPlus_TC_QuickJitForLoops` set, and setting `COMPlus_TC_OnStackReplacement` did indeed solve this. Thanks.

I did try splitting the loop, but couldn't seem to force it out of Tier0. I also looked at setting `AggressiveOptimization` on the method, but it was in f# and found that the f# compiler currently doesn't propagate that attribute to the generated IL, so that was a failure too. 

Anyway, my main concern that this could of been an issue going forward has been alleviated, so thank for that.
